pkg/reconcile/reconcile.go

```diff
@@ -91,3 +93,31 @@ var _ Reconciler = Func(nil)
 
 // Reconcile implements Reconciler.
 func (r Func) Reconcile(o Request) (Result, error) { return r(o) }
+
+// PanicHandler can be called with defer in order to recover from panic in reconcile logic.
+// Additional handlers can also be passed to manage the go routines. The error passed to the
+// PanicHandler is set to the value of panic if not nil.
+// If any resource triggers a panic, the reconciler will recover and modify the raised panic
+// to the normal reconciler error.
+// Example usage:
+/*
+
+import (
+	...
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	)
+
+func (a *ReplicaSetReconciler) Reconcile(req reconcile.Request) (result reconcile.Result, err error) {
+	defer PanicHandler(&err)
+	// reconciler logic here
+
+	...
+*/
+func PanicHandler(err *error, panicHandlers ...func(interface{})) {
```

Comment:
Nit: The names for the helper `PanicHandler()` and the additional handlers that are passed in `panicHandlers` are a bit similar.

How about making the function name more actionable e.g `HandlePanic()` or `HandleCrash()`, and the extra handlers as `additionalHandlers`.
```suggestion
func HandlePanic(err *error, additionalHandlers ...func(interface{})) {
```

Reply:
My suggestion for it would be : 
```suggestion
func HandlePanicWith(err *error, handleOptions ...func(interface{})) {
```